File: 프로그래머스2단계/게임맵최단거리.cpp
#include<vector>
#include<queue>

using namespace std;
vector<vector<int> > Maps;
int rowLen, colLen;
int dx[4] = {0,1,0,-1};
int dy[4] = {-1,0,1,0};
bool visited[101][101]; 

struct Node{
    int x, y, cost;
};

int bfs(){
    queue<Node> q;
    q.push({0,0,1});
    visited[0][0] = true;
    
    while(!q.empty()){
        int x = q.front().x;
        int y = q.front().y;
        int cost = q.front().cost;
        q.pop();
        
        if(x == colLen-1 && y == rowLen-1){
            return cost;
        }
        
        for(int i=0; i < 4 ; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(nx >=0 && ny >= 0 && nx < colLen && ny < rowLen){
                if(!visited[nx][ny] && Maps[nx][ny] == 1){
                    visited[nx][ny] = true;
                    q.push({nx, ny, cost+1});
                }
            }
        }
    }
    return -1;
}

int solution(vector<vector<int> > maps)
{
    Maps = maps;
    rowLen = maps.size();
    colLen = maps[0].size();
    return bfs();
}

File: 조합,순열,BFS,DFS,Graph,DP,Brute Force/게임맵최단거리.java
import java.util.*;

//8월5일 한번더 복습함

class 게임최단거리 {
    int[] dy = {-1,0,1,0};
    int[] dx = {0,1,0,-1};
    boolean[][] visited;
    Queue<Node> q;
    int rowLen, colLen;
    
    public class Node{
        int x;
        int y;
        int cost;
        
        public Node(int x, int y, int cost){
            this.x = x;
            this.y = y;
            this.cost = cost;
        }
    }
    
    public int solution(int[][] maps) {
        rowLen = maps.length;
        colLen = maps[0].length;
        visited = new boolean[rowLen][colLen];
        
        return bfs(0,0,maps);
    }
    
    public int bfs(int x, int y, int[][] maps){
        q = new LinkedList<>();
        q.offer(new Node(x,y,1));
        
        while(!q.isEmpty()){
            Node node = q.poll();
            if(node.x == colLen-1 && node.y == rowLen-1){
                return node.cost;
            }
            
            for(int i = 0; i < 4; i++){
                int nx = node.x + dx[i];
                int ny = node.y + dy[i];
                if(nx >= 0 && ny >= 0 && nx < colLen && ny < rowLen){
                    if(!visited[nx][ny] && maps[nx][ny] == 1){
                        visited[nx][y] = true;
                        q.offer(new Node(nx, ny, node.cost+1));
                    }
                }
            }
            
        }
        return -1;
    }
}
